kernel/src/hil/time.rs

```diff
@@ -714,6 +714,95 @@ impl PartialEq for Ticks64 {
 
 impl Eq for Ticks64 {}
 
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum DayOfWeek {
+    Sunday,
+    Monday,
+    Tuesday,
+    Wednesday,
+    Thursday,
+    Friday,
+    Saturday,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Month {
+    January,
+    February,
+    March,
+    April,
+    May,
+    June,
+    July,
+    August,
+    September,
+    October,
+    November,
+    December,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct DateTime {
+    pub year: u16,
+    pub month: Month,
+    pub day: u8,
+    pub day_of_week: DayOfWeek,
+    pub hour: u8,
+    pub minute: u8,
+    pub seconds: u8,
+}
```

Comment:
Time is complex; I have multiple questions:

1. Does all RTC hardware keep track of year/month/day/day_of_week/hour/minute/seconds separately like this? I would expect that some RTCs would track seconds, and would be used to track time since an epoch or Unix time.
2. What timezone is a `DateTime` in? If the answer is "that's application-specific", then what if multiple processes want to interact with the RTC syscall API?
3. How are leap seconds handled? Do we ignore them, and assume that the RTC time will be corrected occasionally? I don't think we want to encode a leap second table in Tock.

Reply:
> Time is complex; this definition poses several questions:
> 
> 
> 
> 1. Does all RTC hardware keep track of year/month/day/day_of_week/hour/minute/seconds separately like this? I would expect that some RTCs would track seconds, and would be used to track time since an epoch or Unix time.
> 

You are probably right, but I'm not sure how to proceed here without having the kernel perform a conversion. From the pure HIL point of view, I think that its purpose is to track the time, not just seconds.

A possibility would be to have another UnixRtc HIL that interacts with unix time rtc hardware.

> 2. What timezone is a `DateTime` in? If the answer is "that's application-specific", then what if multiple processes want to interact with the RTC syscall API?
> 

I think this is a userspace problem rather than a kernel one. The RTC driver reports the stored time, just like for a normal computer. It is up to the userspace library or application to convert it to local time.

> 3. How are leap seconds handled? Do we ignore them, and assume that the RTC time will be corrected occasionally? I don't think we want to encode a leap second table in Tock.

I think that the easiest way is to just ignore it for now. Systems will probably sync time with a sever from time to time and this will correct the leap second problem.

Reply:
> > Time is complex; this definition poses several questions:
> > 
> > 1. Does all RTC hardware keep track of year/month/day/day_of_week/hour/minute/seconds separately like this? I would expect that some RTCs would track seconds, and would be used to track time since an epoch or Unix time.
> 
> You are probably right, but I'm not sure how to proceed here without having the kernel perform a conversion. From the pure HIL point of view, I think that its purpose is to track the time, not just seconds.
> 
> A possibility would be to have another UnixRtc HIL that interacts with unix time rtc hardware.
>

I renamed the HIL to `DateTime` so that its name better reflects what it does: it keeps track of the date and time. Underlying chips may implement this in any way the want, such as RTC or other hardware component.

I also moved it to `hil/src/date_time.rs` to separate it from the time HILs.